text/0000-rvalue_static_promotion.md

```diff
@@ -0,0 +1,171 @@
+# Detailed design
+[design]: #detailed-design
+
+Inside a function body's block:
+
+- If a shared reference to a constexpr rvalue is taken. (`&<constexpr>`)
+- And the constexpr does not contain a `UnsafeCell { ... }` constructor.
+- And the constexpr does not contain a const fn call returning a type containing a `UnsafeCell`.
+- Then instead of translating the value into a stack slot, translate
+  it into a static memory location and give the resulting reference a
+  `'static` lifetime.
+
+Likewise,
+
+- If a mutable reference to a constexpr rvalue is taken. (`&mut <constexpr>`)
+- And the constexpr does not contain a `UnsafeCell { ... }` constructor.
+- And the constexpr does not contain a const fn call returning a type containing a `UnsafeCell`.
+- _And the type of the rvalue is zero-sized._
+- Then instead of translating the value into a stack slot, translate
+  it into a static memory location and give the resulting reference a
+  `'static` lifetime.
+
+The `UnsafeCell` restrictions are there to ensure that the promoted value is
+truly immutable behind the reference (Though not technically needed in the zero-sized case, see alternatives below).
+
+The zero-sized restriction for mutable references is there because
+aliasing mutable references are only safe for zero sized types
+(since you never dereference the pointer for them).
+
+Examples:
+
+```rust
+// OK:
+let a: &'static u32 = &32;
+let b: &'static Option<UnsafeCell<u32>> = &None;
+let c: &'static Fn() -> u32 = &|| 42;
+
+let d: &'static mut () = &mut ();
+let e: &'static mut Fn() -> u32 = &mut || 42;
+
+// BAD:
+let f: &'static Option<UnsafeCell<u32>> = &Some(UnsafeCell { data: 32 });
+let g: &'static Cell<u32> = &Cell::new(); // assuming conf fn new()
+```
+
+These rules above should be consistent with the existing rvalue promotions in `const`
+initializer lists:
+
+```rust
+// If this compiles:
+const X: &'static T = &<constexpr foo>;
+
+// Then this should compile as well:
+let x: &'static T = &<constexpr foo>;
+```
+
+## Implementation
+
+The necessary changes in the compiler did already get implemented as
+part of codegen optimizations (emitting references-to or memcopies-from values in static memory instead of embedding them in the code).
+
+All that is left do do is "throw the switch" for the new lifetime semantic
+by removing these lines:
+https://github.com/rust-lang/rust/blob/29ea4eef9fa6e36f40bc1f31eb1e56bf5941ee72/src/librustc/middle/mem_categorization.rs#L801-L807
+
+(And of course fixing any fallout/bitrot that might have happened, adding tests, etc.)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+One more feature with seemingly ad-hoc rules to complicate the language...
+
+# Alternatives
+[alternatives]: #alternatives
+
+There are two ways this could be taken further with zero-sized types:
+
+1. Remove the `UnsafeCell` restriction if the type of the rvalue is zero-sized.
+2. The above, but also remove the __constexpr__ restriction, applying to any zero-sized rvalue instead.
```

Comment:
Currently types with destructors are not zero-sized but that will change and I agree - there is no reason not to be conservative, for now, IMHO.

For the specific case of destructors, remember that the following are more or less equal:
```rust
let x = &EmptyDestructible;
...
```
```rust
let temp = EmptyDestructible;
{
    let x = &temp;
    ...
}
Drop::drop(&mut temp)
```

Thus, rvalue drops can be considered "mutable accesses".
And even if the destructor doesn't have any bytes to actually mutate, it has to run.
So you would be changing behavior if you remove the destructor call, and otherwise you kind of make a mess out of borrow semantics by allowing the resulting `&'static EmptyDestructible` to escape the scope its destructor ran in.

In the end, no UB can actually occur, but it still seems troubling to do this dance around destructors.